testing/test_service_assertions.py

```diff
@@ -0,0 +1,231 @@
+        service = context.get_service(BaseService)
+        return data_container, context
+
+    def _fit_data_container(self, data_container: DataContainer, context: ExecutionContext):
+        service: BaseService = context.get_service(BaseService)
+        service.service_method(data_container.data_inputs)
+        return self
+
+    def _transform_data_container(self, data_container: DataContainer, context: ExecutionContext):
+        service: BaseService = context.get_service(BaseService)
+        service.service_method(data_container.data_inputs)
+        return data_container
+
+
+class RegisterServiceDynamically(ForceHandleMixin, Identity):
+    def __init__(self):
+        Identity.__init__(self)
+        ForceHandleMixin.__init__(self)
+
+    def _will_process(self, data_container: DataContainer, context: ExecutionContext):
+        # data_container, context = ForceHandleMixin._will_process(data_container, context)
+        context.register_service(BaseService, SomeService())
+        return data_container, context
+
+    def _transform_data_container(self, data_container: DataContainer, context: ExecutionContext):
+        return data_container
+
+
+def test_step_with_context_should_only_save_wrapped_step(tmpdir):
+    context = ExecutionContext(root=tmpdir)
+    service = SomeService()
+    context.set_service_locator({BaseService: service})
+    p = Pipeline([
+        SomeStep().assert_has_services(BaseService)
+    ]).with_context(context=context)
+
+    p.save(context, full_dump=True)
+
+    p: Pipeline = ExecutionContext(root=tmpdir).load(os.path.join('StepWithContext', 'Pipeline'))
+    assert isinstance(p, Pipeline)
+
+
+def test_with_context_should_inject_dependencies_properly(tmpdir):
+    data_inputs = np.array([0, 1, 2, 3])
+    context = ExecutionContext(root=tmpdir)
+    service = SomeService()
+    context.set_service_locator({BaseService: service})
+    p = Pipeline([
+        SomeStep().assert_has_services(BaseService),
+        SomeStep().assert_has_services_at_execution(BaseService)
+    ]).with_context(context=context)
+
+    p.transform(data_inputs=data_inputs)
+
+    assert np.array_equal(service.data, data_inputs)
+
+
+def test_with_context_should_fail_at_init_when_services_are_missing(tmpdir):
+    context = ExecutionContext(root=tmpdir)
+    p = Pipeline([
+        SomeStep().assert_has_services(BaseService)
+    ]).with_context(context=context)
+
+    data_inputs = np.array([0, 1, 2, 3])
+    with pytest.raises(AssertionError) as exception_info:
+        p.transform(data_inputs=data_inputs)
+
+    assert 'BaseService dependency missing' in exception_info.value.args[0]
+
+
+def test_localassert_should_assert_dependencies_properly_at_exec(tmpdir):
+    data_inputs = np.array([0, 1, 2, 3])
+    context = ExecutionContext(root=tmpdir)
+    p = Pipeline([
+        RegisterServiceDynamically(),
+        SomeStep().assert_has_services_at_execution(BaseService)
```

Comment:
Optional: 
I suggest also coding a negative version of this test : testing if a global service assertion will effectively raise and AssertionError if it is placed after the dynamic registration.

Reply:
Good call! I modified "test_auto_ml_should_fail_at_init_when_services_are_missing" to include a dynamical registering of service as to make sure that the assertion raised is the global one and not the local one.

Reply:
Re-writing this test made me notice that AutoML did not implement _HasChildrenMixin which broke the usage of apply function with it!